File: daily_python/3306_Count_of_Substrings_Containing_Every_Vowel_and_K_Consonants_II.py
import os
import sys
from typing import List

class Solution:
    def _isVowel(self, c: chr) -> bool:
        return c in "aeiou"
    def _valid(self, count: dict) -> bool:
        return all(vowel in count and count[vowel] > 0 for vowel in "aeiou")
    def countOfSubstrings(self, word: str, k: int) -> int:
        n = len(word)
        next_consonant = [n] * n
        idx = n
        for i in range(n-1, -1, -1):
            next_consonant[i] = idx
            if not self._isVowel(word[i]):
                idx = i
        start = 0
        end = 0
        consonant_count = 0
        count = {}
        ans = 0
        while end < n:
            added = word[end]
            if self._isVowel(added):
                count[added] = count.get(added, 0) + 1
            else:
                consonant_count += 1
            while start < end and self._valid(count) and consonant_count >= k:
                if consonant_count == k:
                    ans += next_consonant[end] - end
                delete = word[start]
                if self._isVowel(delete):
                    count[delete] -= 1
                else:
                    consonant_count -= 1
                start += 1
            end += 1
        return ans
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            word = lines[0].split('=')[1].strip()[1:-2]
            k = int(lines[1].split('=')[1].strip()[:-1])
            dataset.append((word, k))
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for word, k in dataset:
        results.append(solution.countOfSubstrings(word, k))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')

File: daily/2025/3306_Count_of_Substrings_Containing_Every_Vowel_and_K_Consonants_II.cpp
#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>

using namespace std;

class Solution {
public:
    long long countOfSubstrings(string word, int k) {
        return morethank(word, k) - morethank(word, k+1);
    }
    long long morethank(string word, int k){
        int n = word.size();
        int start = 0, end = 0;
        int consonant = 0;
        unordered_map<char,int> vowel;
        long long ans = 0;
        while(end < n){
            char added = word[end];
            if (isvowel(added)) vowel[added]++;
            else consonant++;
            while(vowel.size() == 5 && consonant >= k){
                ans += n - end;
                char del = word[start];
                if (isvowel(del)){
                    vowel[del]--;
                    if (vowel[del] == 0) vowel.erase(del);
                }
                else consonant--;
                start++;
            }
            end++;
        }
        return ans;
    }
    bool isvowel(char c){return c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u';}
};

int main(){
    string word = "aeioqq";
    int k = 1;
    // string word = "aeiou";
    // int k = 0;
    // string word = "ieaouqqieaouqq";
    // int k = 1;
    // string word = "iqeaouqi";
    // int k = 2;
    Solution S;
    cout << S.countOfSubstrings(word, k) << endl;
    return 0;
}